#include "resultswindow.h"
#include "ui_resultswindow.h"
#include <QTableWidgetItem>
#include <QHeaderView>
#include <algorithm>
#include <QDebug>

ResultsWindow::ResultsWindow(const QStringList &players, const QVector<int> &scores, QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::ResultsWindow)
{
    ui->setupUi(this);

    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Å—Ç–∏–ª—å –æ–∫–Ω–∞
    this->setStyleSheet("QDialog { background-color: #1a1a2e; font-family: 'Segoe UI', Arial, sans-serif; }");

    setWindowTitle("üèÜ –†–µ–∑—É–ª—å—Ç–∞—Ç—ã –∏–≥—Ä—ã");
    setMinimumSize(800, 600);

    // –ü–æ–ª—É—á–∞–µ–º –æ—Ç—Å–æ—Ä—Ç–∏—Ä–æ–≤–∞–Ω–Ω—ã–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç—ã
    QVector<QPair<QString, int>> sortedResults = getSortedResults(players, scores);

    // –ù–∞—Å—Ç—Ä–∞–∏–≤–∞–µ–º —Ç–∞–±–ª–∏—Ü—É —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
    setupResultsTable(sortedResults);

    // –ü–æ–¥–∫–ª—é—á–∞–µ–º –∫–Ω–æ–ø–∫—É –∑–∞–∫—Ä—ã—Ç–∏—è
    connect(ui->closeButton, &QPushButton::clicked, this, &ResultsWindow::on_closeButton_clicked);
}

QVector<QPair<QString, int>> ResultsWindow::getSortedResults(const QStringList &players, const QVector<int> &scores) const
{
    QVector<QPair<QString, int>> results;

    // –°–æ–±–∏—Ä–∞–µ–º –ø–∞—Ä—ã –∏–≥—Ä–æ–∫-–æ—á–∫–∏
    for (int i = 0; i < players.size(); ++i) {
        results.append(qMakePair(players[i], scores[i]));
    }

    // –°–æ—Ä—Ç–∏—Ä—É–µ–º –ø–æ —É–±—ã–≤–∞–Ω–∏—é –æ—á–∫–æ–≤
    std::sort(results.begin(), results.end(),
              [](const QPair<QString, int> &a, const QPair<QString, int> &b) {
                  return a.second > b.second;
              });

    return results;
}

QString ResultsWindow::getPlaceText(int place) const
{
    switch (place) {
    case 1: return "ü•á 1-–µ –º–µ—Å—Ç–æ";
    case 2: return "ü•à 2-–µ –º–µ—Å—Ç–æ";
    case 3: return "ü•â 3-–µ –º–µ—Å—Ç–æ";
    default: return QString("üèÖ %1-–µ –º–µ—Å—Ç–æ").arg(place);
    }
}

QString ResultsWindow::getPlaceStyle(int place) const
{
    switch (place) {
    case 1:
        return R"(
            QLabel {
                font-size: 22px;
                font-weight: bold;
                color: #ffd700;
                padding: 15px;
                background: qlineargradient(x1:0, y1:0, x2:1, y2:0,
                    stop:0 #ffd700, stop:0.3 #fffacd);
                border-radius: 10px;
                border: 3px solid #ffd700;
            }
        )";
    case 2:
        return R"(
            QLabel {
                font-size: 20px;
                font-weight: bold;
                color: #c0c0c0;
                padding: 15px;
                background: qlineargradient(x1:0, y1:0, x2:1, y2:0,
                    stop:0 #c0c0c0, stop:0.3 #f5f5f5);
                border-radius: 10px;
                border: 3px solid #c0c0c0;
            }
        )";
    case 3:
        return R"(
            QLabel {
                font-size: 18px;
                font-weight: bold;
                color: #cd7f32;
                padding: 15px;
                background: qlineargradient(x1:0, y1:0, x2:1, y2:0,
                    stop:0 #cd7f32, stop:0.3 #f4d3a1);
                border-radius: 10px;
                border: 3px solid #cd7f32;
            }
        )";
    default:
        return R"(
            QLabel {
                font-size: 16px;
                font-weight: bold;
                color: #3498db;
                padding: 15px;
                background-color: #16213e;
                border-radius: 10px;
                border: 2px solid #3498db;
            }
        )";
    }
}

void ResultsWindow::setupResultsTable(const QVector<QPair<QString, int>> &sortedResults)
{
    ui->tableWidget->setRowCount(sortedResults.size());
    ui->tableWidget->setColumnCount(3);  // –ò–∑–º–µ–Ω–µ–Ω–æ —Å 4 –Ω–∞ 3 (—É–±—Ä–∞–ª–∏ —Å—Ç–æ–ª–±–µ—Ü "–î–æ—Å—Ç–∏–∂–µ–Ω–∏–µ")

    QStringList headers;
    headers << "–ú–µ—Å—Ç–æ" << "–ò–≥—Ä–æ–∫" << "–û—á–∫–∏";  // –£–±—Ä–∞–ª–∏ "–î–æ—Å—Ç–∏–∂–µ–Ω–∏–µ"
    ui->tableWidget->setHorizontalHeaderLabels(headers);

    // –ù–∞—Å—Ç—Ä–æ–π–∫–∞ —Ç–∞–±–ª–∏—Ü—ã
    ui->tableWidget->horizontalHeader()->setStyleSheet(R"(
        QHeaderView::section {
            background-color: #0f3460;
            color: white;
            font-weight: bold;
            font-size: 14px;
            padding: 10px;
            border: 1px solid #2d4059;
        }
    )");

    ui->tableWidget->verticalHeader()->setVisible(false);
    ui->tableWidget->setShowGrid(false);
    ui->tableWidget->setAlternatingRowColors(true);
    ui->tableWidget->setStyleSheet(R"(
        QTableWidget {
            background-color: #16213e;
            color: white;
            font-size: 14px;
            border: 2px solid #2d4059;
            border-radius: 10px;
            gridline-color: #2d4059;
        }
        QTableWidget::item {
            padding: 15px;
            border-bottom: 1px solid #2d4059;
        }
        QTableWidget::item:selected {
            background-color: #3498db;
            color: white;
        }
        QScrollBar:vertical {
            border: none;
            background: #1a1a2e;
            width: 12px;
            border-radius: 6px;
        }
        QScrollBar::handle:vertical {
            background: #2d4059;
            border-radius: 6px;
            min-height: 30px;
        }
    )");

    // –ó–∞–ø–æ–ª–Ω—è–µ–º —Ç–∞–±–ª–∏—Ü—É
    for (int i = 0; i < sortedResults.size(); ++i) {
        int place = i + 1;
        const auto &player = sortedResults[i].first;
        int score = sortedResults[i].second;

        // –ú–µ—Å—Ç–æ
        QTableWidgetItem *placeItem = new QTableWidgetItem(getPlaceText(place));
        placeItem->setTextAlignment(Qt::AlignCenter);
        placeItem->setFlags(placeItem->flags() ^ Qt::ItemIsEditable);

        // –ò–≥—Ä–æ–∫
        QTableWidgetItem *playerItem = new QTableWidgetItem(player);
        playerItem->setTextAlignment(Qt::AlignCenter);
        playerItem->setFlags(playerItem->flags() ^ Qt::ItemIsEditable);

        // –û—á–∫–∏
        QTableWidgetItem *scoreItem = new QTableWidgetItem(QString::number(score));
        scoreItem->setTextAlignment(Qt::AlignCenter);
        scoreItem->setFlags(scoreItem->flags() ^ Qt::ItemIsEditable);

        // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Å—Ç–∏–ª–∏ –¥–ª—è —Å—Ç—Ä–æ–∫
        QFont font;
        font.setBold(true);
        placeItem->setFont(font);

        if (place == 1) {
            placeItem->setForeground(QColor(255, 215, 0)); // –ó–æ–ª–æ—Ç–æ–π
            playerItem->setForeground(QColor(255, 215, 0));
            scoreItem->setForeground(QColor(255, 215, 0));
        } else if (place == 2) {
            placeItem->setForeground(QColor(192, 192, 192)); // –°–µ—Ä–µ–±—Ä—è–Ω—ã–π
            playerItem->setForeground(QColor(192, 192, 192));
            scoreItem->setForeground(QColor(192, 192, 192));
        } else if (place == 3) {
            placeItem->setForeground(QColor(205, 127, 50)); // –ë—Ä–æ–Ω–∑–æ–≤—ã–π
            playerItem->setForeground(QColor(205, 127, 50));
            scoreItem->setForeground(QColor(205, 127, 50));
        }

        ui->tableWidget->setItem(i, 0, placeItem);
        ui->tableWidget->setItem(i, 1, playerItem);
        ui->tableWidget->setItem(i, 2, scoreItem);
    }

    // –†–µ—Å–∞–π–∑–∏–º –∫–æ–ª–æ–Ω–∫–∏
    ui->tableWidget->resizeColumnsToContents();
    ui->tableWidget->horizontalHeader()->setStretchLastSection(true);

    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –≤—ã—Å–æ—Ç—É —Å—Ç—Ä–æ–∫
    for (int i = 0; i < ui->tableWidget->rowCount(); ++i) {
        ui->tableWidget->setRowHeight(i, 60);
    }

    // –û–±—â–∞—è —Å—É–º–º–∞ –æ—á–∫–æ–≤
    int totalScore = 0;
    for (const auto &result : sortedResults) {
        totalScore += result.second;
    }
    ui->totalScoreLabel->setText(QString("–û–±—â–∞—è —Å—É–º–º–∞ –æ—á–∫–æ–≤: <span style='color:#ffd700; font-size: 18pt;'>%1</span>").arg(totalScore));

    // –ü–æ–±–µ–¥–∏—Ç–µ–ª—å
    if (!sortedResults.isEmpty()) {
        QString winner = sortedResults[0].first;
        int winnerScore = sortedResults[0].second;
        ui->winnerLabel->setText(QString("–ü–æ–±–µ–¥–∏—Ç–µ–ª—å: <span style='color:#ffd700; font-size: 20pt; font-weight: bold;'>%1</span> —Å <span style='color:#2ecc71; font-size: 18pt;'>%2 –æ—á–∫–∞–º–∏</span>!").arg(winner).arg(winnerScore));
    }
}

ResultsWindow::~ResultsWindow()
{
    delete ui;
}

void ResultsWindow::on_closeButton_clicked()
{
    emit returnToMainMenu();  // –ò—Å–ø—É—Å–∫–∞–µ–º —Å–∏–≥–Ω–∞–ª
    this->accept();           // –ó–∞–∫—Ä—ã–≤–∞–µ–º –æ–∫–Ω–æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
}
